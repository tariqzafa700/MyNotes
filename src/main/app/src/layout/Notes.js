import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

const styles = theme => ({
	  container: {
	    display: 'flex',
	    flexWrap: 'wrap',
	    flexDirection: 'column',
	    justifyContent: "center",
	    alignItems: 'center'
	  },
	  title: {
	    marginLeft: theme.spacing.unit,
	    marginRight: theme.spacing.unit,
	    width: 300,
	    borderStyle: 'outset',
	    borderWidth: '2px',
	    borderRadius: '2px'
	  },
	  content: {
		marginLeft: theme.spacing.unit,
	    marginRight: theme.spacing.unit,
	    width: 300,
	    borderStyle: 'outset',
	    borderWidth: '2px',
	    borderRadius: '2px'
	  },
	  button: {
	    marginTop: 20,
	    borderRadius: '4px'
	  }
});

export class Notes extends Component {
	constructor(props) {
	    super(props);
	    this.state = {
	    		title: "",
	    		content: "",
	    		savedEntries: []
	    }
	}
	
	componentDidMount = () => {
		fetch("/notes", {
	        method: "GET",
	        headers: {
	          'Accept': 'application/json',
	        },
	        credentials: "same-origin"
    	}).then(function(response) {
	        return response.json();
	    }).then(text => {
	        this.setState({savedEntries: text})
	    }).catch(error => {});
	}
	
	handleSubmit = evt => {
		evt.preventDefault();
		const { title, content } = this.state;
        if(title.trim() === "" || content.trim() === "")
        	return;
		fetch("/notes", {
	        method: "POST",
	        headers: {
	          'Accept': 'application/json',
	          'Content-Type': 'application/json'
	        },
	        credentials: "same-origin",
	        body: JSON.stringify({title: title, content: content})
	    }).then(function(response) {
	        return response.json();
	    }).then((text) => {
	        return fetch("/notes", {
		        method: "GET",
		        headers: {
		          'Accept': 'application/json',
		        },
		        credentials: "same-origin"
	    	});
	    }).then(function(response) {
	        return response.json();
	    }).then(text => {
	        this.setState({title: "", content: "", savedEntries: text})
	    }).catch(error => {});
	}
	
	handleChange = name => event => {
		this.setState({
			[name] : event.target.value
		});
	}
	
	notesList = () => {
		var notes = [];
		
		this.state.savedEntries.length > 0? this.state.savedEntries.slice().reverse().forEach(entry => notes.push(<ListItem>
	        <ListItemText primary={entry.title} secondary = {entry.content}></ListItemText>
	    </ListItem>)) : <div/>
		return (
				<List>
				    {notes}
				</List>
		);
	}
	
	render() {
		const { classes } = this.props;
		const { title, content } = this.state;
		return (
				<div>
					<form className={classes.container} onSubmit={this.handleSubmit}>
					  <TextField
				        label="Title"
				        id="margin-title"
				        className={classes.title}
				        margin="normal"
				        value={title}
					    onChange={this.handleChange("title")}
					    inputProps={{ maxLength: 255 }}
				      />
				      <TextField
				        label="Content"
				        id="margin-normal"
				        className={classes.content}
				        multiline
				        rows = {10}
					    value={content}
					    onChange={this.handleChange("content")}
					    inputProps={{ maxLength: 1024 }}
				      />
					  <Button type="submit" variant="raised" color="primary" className={classes.button}>
			            Create
			          </Button>
				    </form>
				    {this.notesList()}
				</div>
		);
	}
}

export default withStyles(styles)(Notes);

