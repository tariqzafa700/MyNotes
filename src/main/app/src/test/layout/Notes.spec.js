import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, render } from 'enzyme';
import fetchMock from 'fetch-mock';
import { createShallow } from '@material-ui/core/test-utils';
import toJson from 'enzyme-to-json';
import Notes from "../../layout/Notes";

describe('Notes Component', () => {
	it('renders correctly', () => {
	  const tree = renderer.create(
	    <Notes />
	  ).toJSON();
	 
	  expect(tree).toMatchSnapshot();
	});
	/*There is some issue with enzyme 3.0 because of which shallow renderer is not working properly.
	 * But below is the test that checks the button element in the renderer and simulates a submit event.
	 * It is expected that fetch mock call is triggered*/
	/*
	it('submits data on create button', () => {
		Enzyme.configure({ adapter: new Adapter() })
		  const tree = shallow(
				  <Notes />
				  );
		  const headers = {
			        "Content-Type": "application/json"
			      };
		  const message = JSON.stringify({
		        title: "title",
		        content: "content"
		      });
		  fetchMock
	        .post("/notes", {
	          status: 201,
	          headers,
	          body: { message }
	        })
	        .catch(unmatchedUrl => {
	          return realFetch(unmatchedUrl);
	        });
          tree.find("button").simulate("submit");
          expect(fetchMoc.called("/notes")).toEqual(true);
	});*/
});