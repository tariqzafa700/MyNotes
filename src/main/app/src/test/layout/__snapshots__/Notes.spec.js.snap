// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes Component renders correctly 1`] = `"<Notes classes={{...}} />"`;
